stdlib/REPL/src/REPL.jl

```diff
@@ -1249,23 +1250,44 @@ function setup_interface(
                             buf = copy(LineEdit.buffer(s))
                             transition(s, pkg_mode) do
                                 LineEdit.state(s, pkg_mode).input_buffer = buf
-                            end
-                            if !isempty(s)
-                                @invokelatest(LineEdit.check_for_hint(s)) && @invokelatest(LineEdit.refresh_line(s))
+                                if !isempty(s)
+                                    @invokelatest(LineEdit.check_for_hint(s)) && @invokelatest(LineEdit.refresh_line(s))
+                                end
+                                if return_pressed # if return was pressed run what was entered
+                                    LineEdit.commit_line(s)
+                                    terminal = LineEdit.terminal(s)
+                                    @invokelatest LineEdit.mode(s).on_done(s, LineEdit.buffer(s), true)
+                                    LineEdit.push_undo(s)
+                                end
                             end
                             transition_finished = true
                         end
                     end
+                    REPLExt
                 end
-                Base.errormonitor(t_replswitch)
                 # while loading just accept all keys, no keymap functionality
-                while !istaskdone(t_replswitch)
-                    # wait but only take if task is still running
-                    peek(stdin, Char)
-                    @lock iolock begin
-                        transition_finished || edit_insert(s, read(stdin, Char))
+                # use @async here to keep this on the same thread as the REPL. With @spawn both tasks
+                # can get stuck waiting for each other
+                t_stdincopy = @async while !istaskdone(t_replswitch)
+                    try
+                        peek(stdin, Char)
+                    catch
+                        # ignore interrupt
+                    else
```

Comment:
This is legal? TIL 

It's not documented in `?try` or `?catch` (but neither is `finally`).

Reply:
https://github.com/JuliaLang/julia/issues/54686